File: app/src/main/java/com/dhpcs/liquidity/fragment/RestoreIdentityDialogFragment.java
package com.dhpcs.liquidity.fragment;

import android.app.Dialog;
import android.content.Context;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatDialogFragment;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.TextView;

import com.dhpcs.liquidity.R;
import com.dhpcs.liquidity.activity.BoardGameActivity;
import com.dhpcs.liquidity.boardgame.BoardGame.Identity;
import com.dhpcs.liquidity.boardgame.BoardGame.IdentityWithBalance;
import com.dhpcs.liquidity.model.MemberId;
import com.dhpcs.liquidity.model.ZoneId;
import com.dhpcs.liquidity.view.Identicon;

import java.util.ArrayList;

import scala.collection.JavaConversions;

public class RestoreIdentityDialogFragment extends AppCompatDialogFragment {

    public interface Listener {

        void onIdentityRestorationRequested(Identity identity);

    }

    public static final String TAG = "restore_identity_dialog_fragment";

    private static class IdentitiesAdapter extends ArrayAdapter<IdentityWithBalance> {

        IdentitiesAdapter(Context context) {
            super(context,
                    R.layout.linearlayout_identity,
                    R.id.textview_name);
        }

        @NonNull
        @Override
        public View getView(int position, View convertView, @NonNull ViewGroup parent) {
            View view = super.getView(position, convertView, parent);

            IdentityWithBalance identity = getItem(position);

            Identicon identiconId = view.findViewById(R.id.identicon_id);
            TextView textViewName = view.findViewById(R.id.textview_name);
            TextView textViewBalance = view.findViewById(R.id.textview_balance);

            assert identity != null;
            ZoneId zoneId = identity.zoneId();
            MemberId memberId = identity.member().id();
            String name = BoardGameActivity.formatNullable(
                    getContext(),
                    identity.member().name()
            );
            String balance = BoardGameActivity.formatCurrencyValue(
                    getContext(),
                    identity.balanceWithCurrency()._2(),
                    identity.balanceWithCurrency()._1()
            );

            identiconId.show(zoneId, memberId);
            textViewName.setText(name);
            textViewBalance.setText(balance);

            return view;
        }

    }

    private static final String ARG_IDENTITIES = "identities";

    public static RestoreIdentityDialogFragment newInstance(
            scala.collection.Iterable<IdentityWithBalance> identities) {
        RestoreIdentityDialogFragment restoreIdentityDialogFragment =
                new RestoreIdentityDialogFragment();
        Bundle args = new Bundle();
        args.putSerializable(
                ARG_IDENTITIES,
                new ArrayList<>(
                        JavaConversions.bufferAsJavaList(
                                identities.toBuffer()
                        )
                )
        );
        restoreIdentityDialogFragment.setArguments(args);
        return restoreIdentityDialogFragment;
    }

    private Listener listener;

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        listener = (Listener) context;
    }

    @NonNull
    @Override
    @SuppressWarnings("unchecked")
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        final ArrayAdapter<IdentityWithBalance> identitiesAdapter = new IdentitiesAdapter(
                getActivity()
        );
        ArrayList<IdentityWithBalance> identities =
                (ArrayList<IdentityWithBalance>) getArguments().getSerializable(ARG_IDENTITIES);
        assert identities != null;
        identitiesAdapter.addAll(identities);
        identitiesAdapter.sort(BoardGameActivity.playerComparator(getActivity()));
        return new AlertDialog.Builder(getActivity())
                .setTitle(R.string.choose_identity_to_restore)
                .setAdapter(identitiesAdapter, (dialog, which) -> {
                    if (listener != null) {
                        listener.onIdentityRestorationRequested(
                                identitiesAdapter.getItem(which)
                        );
                    }
                })
                .create();
    }

    @Override
    public void onDetach() {
        super.onDetach();
        listener = null;
    }

}

File: app/src/main/java/com/dhpcs/liquidity/fragment/RestoreIdentityDialogFragment.kt
package com.dhpcs.liquidity.fragment

import android.app.Dialog
import android.content.Context
import android.os.Bundle
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.TextView
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatDialogFragment
import androidx.lifecycle.ViewModelProviders
import com.dhpcs.liquidity.BoardGame
import com.dhpcs.liquidity.LiquidityApplication
import com.dhpcs.liquidity.R
import com.dhpcs.liquidity.activity.MainActivity
import com.dhpcs.liquidity.view.Identicon

class RestoreIdentityDialogFragment : AppCompatDialogFragment() {

    companion object {

        private class IdentitiesAdapter internal constructor(context: Context
        ) : ArrayAdapter<BoardGame.Companion.Identity>(context,
                R.layout.linearlayout_identity,
                R.id.textview_name
        ) {

            override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
                val view = super.getView(position, convertView, parent)

                val identity = getItem(position)

                val identiconId = view.findViewById<Identicon>(R.id.identicon_id)
                val textViewName = view.findViewById<TextView>(R.id.textview_name)
                val textViewBalance = view.findViewById<TextView>(R.id.textview_balance)

                val zoneId = identity!!.zoneId
                val memberId = identity.memberId
                val name = LiquidityApplication.formatNullable(context, identity.name)
                val balance = LiquidityApplication.formatCurrencyValue(
                        context,
                        identity.currency,
                        identity.balance
                )

                identiconId.show(zoneId, memberId)
                textViewName.text = name
                textViewBalance.text = balance

                return view
            }

        }

        const val TAG = "restore_identity_dialog_fragment"

        fun newInstance(): RestoreIdentityDialogFragment = RestoreIdentityDialogFragment()

    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {

        val model = ViewModelProviders.of(requireActivity())
                .get(MainActivity.Companion.BoardGameModel::class.java)
        val identities = model.boardGame.hiddenIdentities
        val identitiesAdapter = IdentitiesAdapter(requireContext())
        identitiesAdapter.addAll(identities.values)
        identitiesAdapter.sort(LiquidityApplication.identityComparator(requireContext()))

        return AlertDialog.Builder(requireContext())
                .setTitle(R.string.choose_identity_to_restore)
                .setAdapter(identitiesAdapter) { _, which ->
                    model.execCommand(
                            model.boardGame.restoreIdentity(identitiesAdapter.getItem(which)!!)
                    ) {
                        getString(
                                R.string.restore_identity_error_format_string,
                                LiquidityApplication.formatNullable(
                                        requireContext(),
                                        identitiesAdapter.getItem(which)!!.name
                                )
                        )
                    }
                }
                .create()
    }

}
